crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -1,258 +1,148 @@
-// Copyright 2019 Google LLC
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     https://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-// modified by Bevy contributors
-
 extern crate proc_macro;
 
-use std::borrow::Cow;
-
 use find_crate::{Dependencies, Manifest};
 use proc_macro::TokenStream;
 use proc_macro2::{Span, TokenStream as TokenStream2};
-use quote::quote;
+use quote::{format_ident, quote};
 use syn::{
-    parse::ParseStream, parse_macro_input, punctuated::Punctuated, Data, DataStruct, DeriveInput,
-    Error, Field, Fields, GenericParam, Ident, Index, Lifetime, Path, Result, Token,
+    parse::{Parse, ParseStream},
+    parse_macro_input,
+    punctuated::Punctuated,
+    token::Comma,
+    Data, DataStruct, DeriveInput, Field, Fields, GenericParam, Ident, Index, Lifetime, LitInt,
+    Path, Result, Token,
 };
 
-/// Implement `Bundle` for a monomorphic struct
-///
-/// Using derived `Bundle` impls improves spawn performance and can be convenient when combined with
-/// other derives like `serde::Deserialize`.
-#[proc_macro_derive(Bundle)]
-pub fn derive_bundle(input: TokenStream) -> TokenStream {
-    let input = parse_macro_input!(input as DeriveInput);
-    match derive_bundle_(input) {
-        Ok(ts) => ts,
-        Err(e) => e.to_compile_error(),
-    }
-    .into()
+struct AllTuples {
+    macro_ident: Ident,
+    start: usize,
+    end: usize,
+    idents: Vec<Ident>,
 }
 
-#[allow(clippy::cognitive_complexity)]
-fn derive_bundle_(input: DeriveInput) -> Result<TokenStream2> {
-    let ident = input.ident;
-    let data = match input.data {
-        syn::Data::Struct(s) => s,
-        _ => {
-            return Err(Error::new_spanned(
-                ident,
-                "derive(Bundle) does not support enums or unions",
-            ))
+impl Parse for AllTuples {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let macro_ident = input.parse::<Ident>()?;
+        input.parse::<Comma>()?;
+        let start = input.parse::<LitInt>()?.base10_parse()?;
+        input.parse::<Comma>()?;
+        let end = input.parse::<LitInt>()?.base10_parse()?;
+        input.parse::<Comma>()?;
+        let mut idents = Vec::new();
+        while let Ok(ident) = input.parse::<Ident>() {
+            idents.push(ident);
+            let _ = input.parse::<Comma>();
```

Comment:
```rs
input.parse::<Comma>().ok()
```
I think this would be better

Reply:
Is it intended that this will parse `foo bar baz` without comma's in between?

Reply:
I think so

Reply:
i opted to make it a bit stricter (now requires commas and at least one item)